src/main/scala/de/unruh/isabelle/pure/Transition.scala

```diff
@@ -0,0 +1,109 @@
+package de.unruh.isabelle.pure
+
+import de.unruh.isabelle.control.Isabelle
+import de.unruh.isabelle.mlvalue.Implicits.listConverter
+import de.unruh.isabelle.mlvalue.{MLValue, MLValueWrapper}
+import Transition.Ops
+
+import scala.language.postfixOps
+
+// Implicits
+import de.unruh.isabelle.mlvalue.Implicits._
+import de.unruh.isabelle.pure.Implicits.theoryConverter
+
+/** Represents a transition between proof states (ML type `Transition`) in the Isabelle process.
+  *
+  * Use `Transition.parseOuterSyntax` to parse Isar source code into into a list of transitions.
+  *
+  * An instance of this class is merely a thin wrapper around an [[mlvalue.MLValue MLValue]].
+  */
+final class Transition private[Transition] (val mlValue: MLValue[Transition])
+    extends MLValueWrapper[Transition] {
+
+  /** Returns the command name (ML function `Toplevel.name_of`).
+   *
+   * Examples: "theory", "section", "lemma", "by", "using", "definition", 'locale", "record", "end".
+   * The name is "<ignored>" if and only if isIgnored is true.
+   * The name is "<malformed>" if and only if isMalformed is true.
+   */
+  def name(implicit isabelle: Isabelle): String = Ops.getName(this).retrieveNow
+  /** Returns the position of this transition's command name. */
+  def position(implicit isabelle: Isabelle): Position = Ops.getPosition(this).retrieveNow
+
+  /** Returns whether this transition introduces a theory ("theory .. begin", ML function `Toplevel.is_init`).
+   *
+   * Init transitions can only be applied to toplevel states in theory mode.
+   */
+  def isInit(implicit isabelle: Isabelle): Boolean = Ops.isInit(this).retrieveNow
+  /** Returns whether the transition is just whitespace, comments, text (ML function `Toplevel.is_ignored`). */
+  def isIgnored(implicit isabelle: Isabelle): Boolean = Ops.isIgnored(this).retrieveNow
+  /** Returns true if the transition is a parsing error  (ML function `Toplevel.is_malformed`). */
+  def isMalformed(implicit isabelle: Isabelle): Boolean = Ops.isMalformed(this).retrieveNow
+
+  /** Execute the transition on a TopLevelState (ML function `Toplevel.command_exception`).
+   *
+   * TODO What does the `int` parameter do? Probably only affects internal Isabelle 'presentation'.
+   */
+  def execute(
+      state: ToplevelState,
+      timeout_in_millis: Option[Int] = None,
```

Comment:
Maybe rename this to just `timeout`, but add in the documentation that it is in milliseconds? (I.e., `@param timeout Timeout in milliseconds`)

Reply:
Ok, I change the type to Option[Duration].

But actually it doesn't seem to have any effect (regardless of the type, even if I just hard-code 2 milliseconds for each transition), at least not on
```scala
  test("execution timeout") {
    val source = raw"lemma foo: True by (sleep 6.0) simp"
    val theory = Theory.mergeTheories("Foo", endTheory=false, List(Theory("Main")))
    val transitions = Transition.parseOuterSyntax(theory, source)
    var state = ToplevelState(theory)
    for ((tr, s) <- transitions) {
      println("tr: " + s)
      state = tr.execute(state, timeout=Some(Duration(2, MILLISECONDS)))
    }
    println(state)
  }
```
Any ideas?

Reply:
I thought maybe the issue is that "sleep" makes the whole Isabelle process sleep instead of just one thread, but now I tried with an actual slow proof and it also doesn't interrupt it (https://stackoverflow.com/questions/43207392/isabelle-hol-proof-by-simp-is-slow-while-value-is-instantaneous)

Reply:
I believe I found the problem. You wrote `Timeout.apply (Time.fromMilliseconds timeout) Toplevel.command_exception int tr st`. Since `Timeout.apply` takes 3 arguments, this means that `Toplevel.command_exception int` is subjected to the timeout (which returns immediately), and then the resulting closure gets `tr` and `st`. However, `Timeout.apply (Time.fromMilliseconds timeout) (Toplevel.command_exception int tr) st` should work. (I didn't try it on the Scala side, but it worked when experimenting directly inside Isabelle/jEdit.)

One thing I noticed was that apparently the granularity of the timeout is actually quite low. I tried (again in Isabelle/jEdit) and even if I had a timeout of a few ms, it would wait one second, Not a big problem but worth mentioning in the docstring.

I like the use of `Duration`, it fits nicely in the Scala world. But then I would change from `Option[Duration]` to just `Duration`, and use `Duration.Inf` for "no timeout".

One final subtlety: I noticed that a timeout of 0 milliseconds means no timeout for Isabelle. This is risky because if I give a duration that will be rounded down (e.g. 1 microsecond), the command might never return (instead of interrupting as early as possible). My suggestion: Instead of `value.toMillus` (in `execute`), use `value.toMillis.max(1)`. This will ensure that the timeout is at least 1ms, meaning it's not ignored. (In reality, it may then of course be 1 second, but that's better than infinite.)

Reply:
Fantastic, it works now, thanks! I added a test as well.

It seems the actual behavior is that it rounds the timeout to exact full seconds, weirdly enough (1,2,3 ms gives 1000-1010ms, while 1500ms gives 2000-2010ms). Maybe it depends on what system clock ML is internally using.

Ok, changed the default to Duration.Inf (thanks a lot for all the pointers, I'm new to scala and ML).

Good point, I added max(1).